### PR TITLE
Add option to run anomaly mode without nan checking

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83481

Fixes https://github.com/pytorch/pytorch/issues/83117

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -3718,6 +3718,23 @@ class Foo(object):
...
+                return torch.tensor(float("nan")).expand(10, 10)
+        def run_fn(a):
+            out = MyFunc.apply(a)
+            return out.sum()
+
+        with torch.autograd.detect_anomaly(check_nan=False):
+            inp = torch.rand(10, 10, requires_grad=True)
+            out = run_fn(inp)
+            out.backward()
+
```
Comment (soulitzer) [inline test/test_autograd.py:3737]: More tests to add:
- Cpp API
- Nested context managers
- more?

**Comment by vadimkantorov** [comment]
could if also have a new argument `check_inf=True` (by default) or check_pos_inf / check_neg_inf for symmetry with nan? (minor)

**Review by albanD — APPROVED** [review]
SGTM
small nit about the naming though

File: torch/_C/__init__.pyi.in [hunk 2]
```diff
@@ -871,8 +871,9 @@ def autocast_increment_nesting() -> _int: ...
 def autocast_decrement_nesting() -> _int: ...
 def is_autocast_cache_enabled() -> _bool: ...
 def set_autocast_cache_enabled(enabled: _bool) -> None: ...
-def set_anomaly_enabled(enabled: _bool) -> None: ...
+def set_anomaly_enabled(enabled: _bool, check_nan: _bool = True) -> None: ...
 def is_anomaly_enabled() -> _bool: ...
+def should_anomaly_check_nan() -> _bool: ...
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/_C/__init__.pyi.in]: Nice, I like that naming much better

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**